pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -0,0 +1,696 @@
+
+        BaseVaultTest.Balances memory balancesBefore = getBalances(sender);
+
+        vm.startPrank(sender);
+        uint256 exactAmountOutDo = router.swapSingleTokenExactIn(
+            pool,
+            tokenA,
+            tokenB,
+            exactAmountIn,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        uint256 exactAmountOutUndo = router.swapSingleTokenExactIn(
+            pool,
+            tokenB,
+            tokenA,
+            exactAmountOutDo,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        vm.stopPrank();
+
+        BaseVaultTest.Balances memory balancesAfter = getBalances(sender);
+
+        assertLe(exactAmountOutUndo, exactAmountIn, "Amount out undo should be <= exactAmountIn");
+
+        _checkUserBalancesAndPoolInvariant(balancesBefore, balancesAfter, 0, 0);
+    }
+
+    function testDoUndoExactInVariableFees__Fuzz(uint256 poolSwapFeePercentage) public {
+        uint256 exactAmountIn = maxSwapAmountTokenA;
+        poolSwapFeePercentage = bound(poolSwapFeePercentage, minPoolSwapFeePercentage, maxPoolSwapFeePercentage);
+
+        vault.manualSetStaticSwapFeePercentage(pool, poolSwapFeePercentage);
+
+        BaseVaultTest.Balances memory balancesBefore = getBalances(sender);
+
+        vm.startPrank(sender);
+        uint256 exactAmountOutDo = router.swapSingleTokenExactIn(
+            pool,
+            tokenA,
+            tokenB,
+            exactAmountIn,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+
+        uint256 feesTokenB = vault.getAggregateSwapFeeAmount(pool, tokenB);
+
+        // In the first swap, the trade was exactAmountIn => exactAmountOutDo + feesTokenB. So, if
+        // there were no fees, trading `exactAmountOutDo + feesTokenB` would get exactAmountIn. Therefore, a swap
+        // with exact_in `exactAmountOutDo + feesTokenB` is comparable to `exactAmountIn`, given that the fees are
+        // known.
+        uint256 exactAmountOutUndo = router.swapSingleTokenExactIn(
+            pool,
+            tokenB,
+            tokenA,
+            exactAmountOutDo + feesTokenB,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        uint256 feesTokenA = vault.getAggregateSwapFeeAmount(pool, tokenA);
+        vm.stopPrank();
+
+        BaseVaultTest.Balances memory balancesAfter = getBalances(sender);
+
+        assertLe(exactAmountOutUndo, exactAmountIn - feesTokenA, "Amount out undo should be <= exactAmountIn");
+
+        _checkUserBalancesAndPoolInvariant(balancesBefore, balancesAfter, feesTokenA, feesTokenB);
+    }
+
+    function testDoUndoExactInVariableFeesAmountInAndLiquidity__Fuzz(
+        uint256 exactAmountIn,
+        uint256 poolSwapFeePercentage,
+        uint256 liquidityTokenA,
+        uint256 liquidityTokenB
+    ) public {
```

Comment:
Ah nice, so this one is fuzzing all at once!

I wonder if we could have a single internal function like this, and then have multiple external ones that fuzz piece by piece (can be done in another PR too)

Reply:
Done!